Add session contents for 03: Benchmarking and Debugging

gabrielmocanu opened the pull request:
WIP

Signed-off-by: Viorel-Gabriel MOCANU (94548) <viorel.mocanu@stud.acs.upb.ro>

gabrielmocanu committed:
03-debug: Add documentation on debugging

Signed-off-by: Viorel-Gabriel MOCANU (94548) <viorel.mocanu@stud.acs.upb.ro>

razvand committed:
03-bench-debug: Review and reformat content

Signed-off-by: Razvan Deaconescu <razvan.deaconescu@cs.pub.ro>

razvand commented on content/en/docs/sessions/03-benchmarking-and-debugging/index.md:
```diff
@@ -0,0 +1,260 @@
+
+## Reminders
+
+At this stage, you should be familiar with the steps of configuring, building and running any application within Unikraft and know the main parts of the architecture.
```
There should be a summary of commands on using `kraft` and `make` to configure, build and run Unikraft images.

razvand commented on content/en/docs/sessions/03-benchmarking-and-debugging/index.md:
````diff
@@ -0,0 +1,260 @@
+
+For the Linux user space target (`linuxu`) simply point GDB to the resulting debug image, for example:
+
+```properties
````
What is `properties`?

gabrielmocanu replied:
I tried to have a general highlights but it seems that docsy isn't supporting highlights in Markdown.
I remove them.

razvand commented on content/en/docs/sessions/03-benchmarking-and-debugging/index.md:
````diff
@@ -0,0 +1,260 @@
+We recommend creating the guest in a paused state (the `-S` option):
+
+```properties
+qemu-system-x86_64 -s -S -cpu host -enable-kvm -m 128 -nodefaults -no-acpi -display none -serial stdio -device isa-debug-exit -kernel build/app-helloworld_kvm-x86_64.dbg -append verbose
````
We should be primarily using `kraft` for this. `kraft` is the main tool we want to use.

skuenzer replied:
Can kraft actually open you a debugger enpoint. I implemented somehting into `qemu-guest` to make it easy for debugging

gabrielmocanu replied:
I didn't find a solution with kraft to open a debugger endpoint.
I added an example of how to use qemu-guest to avoid using so many arguments with qemu-system.

razvand commented on content/en/docs/sessions/03-benchmarking-and-debugging/index.md:
````diff
@@ -0,0 +1,260 @@
+tar remote localhost:1234
+```
+
+You can now run `continue` and debug as you would normally.
````
Add script / scripts to start in debugging mode and to attach with debugging configuration done.

gabrielmocanu replied:
I added one script that is doing all this job.

razvand commented on content/en/docs/sessions/03-benchmarking-and-debugging/index.md:
```diff
@@ -0,0 +1,260 @@
+
+You should be also able to use the debugging file (`build/app-helloworld_xen-x86_64.dbg`) for GDB instead passing the kernel image.
+
+## Tracepoints
```
First detail what the purpose is. Start with why. Then get to tracepoints as a means to that end. Ideally you can make a comparison to some other existing mechanism (maybe kprobes in Linux, or ftrace).

gabrielmocanu replied:
Done.

razvand commented on content/en/docs/sessions/03-benchmarking-and-debugging/index.md:
```diff
@@ -0,0 +1,260 @@
+
+## Benchmarking
+
+## Uniprof
```
Uniprof should only be minimally presented since we don't yet have KVM support.

razvand reviewed (changes requested):
Focus on using `kraft` as much as possible. `kraft` should be the default way to configure, build, run and debug Unikraft instances.

razvanvirtan commented on content/en/docs/sessions/03-benchmarking-and-debugging/index.md:
````diff
@@ -0,0 +1,260 @@
+support/scripts/uk_trace/trace.py list traces.dat
+```
+
+### Creating Tracepoints
````
Maybe moving this section before `Read tracing data` would be more helpful? (I'm thinking that it would be the order of the steps made by the participants while solving tasks: configure, write code, test)

razvanvirtan commented on content/en/docs/sessions/03-benchmarking-and-debugging/index.md:
````diff
@@ -0,0 +1,260 @@
+You can print them by issuing the gdb command `uk trace`:
+
+```properties
+(gdb) uk trace
````
I think listing an output example would be nice, eventually the output generated by the example tracepoint from `Creating Tracepoints`.

gabrielmocanu replied:
I added one tutorial about this and in that section it's an expected output.

gabrielmocanu committed:
03-debug: Improve session contents; add tutorial

Signed-off-by: Viorel-Gabriel MOCANU (94548) <viorel.mocanu@stud.acs.upb.ro>

dragosargint committed:
03-debug: Add work item 02

Signed-off-by: ARGINT DRAGOS IULIAN <dragosargint21@gmail.com>

gabrielmocanu committed:
03-debug: Add work items and solutions

Signed-off-by: Viorel-Gabriel MOCANU (94548) <viorel.mocanu@stud.acs.upb.ro>

gabrielmocanu commented on content/en/docs/sessions/03-benchmarking-and-debugging/index.md:
```diff
@@ -0,0 +1,260 @@
+For Xen the process is slightly more complicated and depends on Xen's `gdbsx` tool.
+First you'll need to make sure you have the tool on your system.
+Here are sample instructions to do that:
+
```
I put a short guide on how to run Xen.

gabrielmocanu committed:
03-debug: Do session updates

Add ELF challenges. Delete part with Uniprof and Ukstore. Add solutions
for 04, 05 and 06 work items.

Signed-off-by: Viorel-Gabriel MOCANU (94548) <viorel.mocanu@stud.acs.upb.ro>

razvanvirtan commented on content/en/docs/sessions/03-benchmarking-and-debugging/index.md:
```diff
@@ -0,0 +1,570 @@
+
+Let's try a new application based on networking, **Nginx**.
+First try to run **Nginx** using kraft.
+After you have successfully run using kraft, try to make a Makefile for **Nginx** and run manually.
```
I think we should warn them about the required order of the libraries in `Makefile` (https://github.com/unikraft/lib-nginx)

gabrielmocanu committed:
03-debug: Review session contents

Mostly fix typos.

Signed-off-by: Viorel-Gabriel MOCANU (94548) <viorel.mocanu@stud.acs.upb.ro>
Co-authored-by: Razvan Virtan <57065879+razvanvirtan@users.noreply.github.com>

dragosargint commented on content/en/docs/sessions/03-benchmarking-and-debugging/index.md:
```diff
@@ -0,0 +1,570 @@
+`p *addr`
+
+You can find more gdb commands [here](https://users.ece.utexas.edu/~adnan/gdb-refcard.pdf)
+
```
```suggestion
Also, if you are unfamiliar with X86_64 calling convention you can read more about it [here] (https://en.wikipedia.org/wiki/X86_calling_conventions)
```

gabrielmocanu commented on content/en/docs/sessions/03-benchmarking-and-debugging/index.md:
```diff
@@ -0,0 +1,570 @@
+### 06. Nginx with or without main? That's the question.
+
+Let's try a new application based on networking, **Nginx**.
+First try to run **Nginx** using kraft.
```
You are right. I was at a past commit where it looks like it was working. I'll wipe the kraft side.

dragosargint commented on content/en/docs/sessions/03-benchmarking-and-debugging/index.md:
```diff
@@ -521,7 +507,7 @@ The output will be similar to this:
 ### 05. Can you trace your own program?
 
-Modify your `Echo-bach Server` application implemented in the (first)[https://usoc21.unikraft.org/docs/sessions/01-baby-steps/#01-echo-back-server] session so that each time the server responds with a message a tracepoint with the corresponding message will be activated.
+Modify your `Echo-bach Server` application implemented in the [first](https://usoc21.unikraft.org/docs/sessions/01-baby-steps/#01-echo-back-server) session so that each time the server responds with a message a tracepoint with the corresponding message will be activated.
```
Shouldn't it be Echo-back?

gabrielmocanu replied:
Yes, I will change.

razvand committed:
03-debug: Review contents

Fix typos, reformat contents.

Signed-off-by: Razvan Deaconescu <razvan.deaconescu@cs.pub.ro>

gabrielmocanu committed:
Fix: Resolve problem with nginx solution. Adjust content.

Signed-off-by: Viorel-Gabriel MOCANU (94548) <viorel.mocanu@stud.acs.upb.ro>

razvand reviewed (approved):
All OK.
